org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerAccess.java

```diff
@@ -0,0 +1,158 @@
+import org.eclipse.lsp4j.InitializeParams;
+import org.eclipse.lsp4j.InitializeResult;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.xtext.ide.serializer.IChangeSerializer;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.resource.IResourceDescriptions;
+import org.eclipse.xtext.resource.impl.ResourceDescriptionsProvider;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.xbase.lib.util.ToStringBuilder;
+
+/**
+ * API for interacting with a running language server.
+ * 
+ * @author Sven Efftinge - Initial contribution and API
+ */
+public interface ILanguageServerAccess {
+	class Context {
+		private final Resource resource;
+
+		private final Document document;
+
+		private final boolean isDocumentOpen;
+
+		private final CancelIndicator cancelChecker;
+
+		public Context(Resource resource, Document document, boolean isDocumentOpen, CancelIndicator cancelChecker) {
+			this.resource = resource;
+			this.document = document;
+			this.isDocumentOpen = isDocumentOpen;
+			this.cancelChecker = cancelChecker;
+		}
+
+		@Override
+		public String toString() {
+			ToStringBuilder b = new ToStringBuilder(this);
+			b.add("resource", resource);
+			b.add("document", document);
+			b.add("isDocumentOpen", isDocumentOpen);
+			b.add("cancelChecker", cancelChecker);
+			return b.toString();
+		}
+
+		public Resource getResource() {
+			return resource;
+		}
+
+		public Document getDocument() {
+			return document;
+		}
+
+		public boolean isDocumentOpen() {
+			return isDocumentOpen;
+		}
+
+		public CancelIndicator getCancelChecker() {
+			return cancelChecker;
+		}
+	}
+
+	class IndexContext {
+		private final IResourceDescriptions index;
+
+		private final CancelIndicator cancelChecker;
+
+		public IndexContext(IResourceDescriptions index, CancelIndicator cancelChecker) {
+			this.index = index;
+			this.cancelChecker = cancelChecker;
+		}
+
+		@Override
+		public String toString() {
+			ToStringBuilder b = new ToStringBuilder(this);
+			b.add("index", index);
+			b.add("cancelChecker", cancelChecker);
+			return b.toString();
+		}
+
+		public IResourceDescriptions getIndex() {
+			return index;
+		}
+
+		public CancelIndicator getCancelChecker() {
+			return cancelChecker;
+		}
+	}
+
+	interface IBuildListener {
+		void afterBuild(List<IResourceDescription.Delta> deltas);
+	}
+
+	/**
+	 * provides read access to a fully resolved resource and Document.
+	 */
+	<T extends Object> CompletableFuture<T> doRead(String uri, Function<ILanguageServerAccess.Context, T> function);
+
+	/**
+	 * Provides read access to the Xtext index.
+	 * 
+	 * @since 2.18
+	 */
+	<T extends Object> CompletableFuture<T> doReadIndex(
+			Function<? super ILanguageServerAccess.IndexContext, ? extends T> function);
+
+	/**
+	 * registers a build listener on the this language server
+	 */
+	void addBuildListener(ILanguageServerAccess.IBuildListener listener);
```

Comment:
Not strictly related: There is no `removeBuildListener`. 
Should we add an empty default method + impl in our code?

Reply:
we would but noone would call

Reply:
I think that's because we don't have any meaningful execute commands in our code base.

Reply:
I would very much prefer this. I've had to remove calls like this before when working on LSP.

People don't call the methods just because they're not there.